externalconn,cloud: add GCP KMS support to external connections

adityamaru opened the pull request:
In https://github.com/cockroachdb/cockroach/pull/84931 we taught the `ExternalStorage` infrastructure to
recognize the `external` URI scheme. In this change we do the same
but for the `KMS` infrastructure.

Concretely, a user is now able to create an external storage
object that represent a gcp KMS URI. This can be done using the
`CREATE EXTERNAL CONNECTION` syntax. The user is then able to point
an operation in CockroachDB such as BACKUP,RESTORE,SHOW BACKUP to that
KMS using an `external` URI. For example:

```
CREATE EXTERNAL CONNECTION backup AS 'nodelocal://1/foo';
CREATE EXTERNAL CONNECTION backupkms AS 'gs:///cmk?AUTH=implict';

BACKUP INTO 'external://foo' WITH kms='external://backupkms';
```

Under the hood, we implement the `ConnectionDetails` interface
for gcp KMS. This allows us to persist a row in the `external_connections`
table when the object is created, and to `Dial` the underlying resource
when the object is being used. The interfaces had to be tweaked slightly
to accomodate for the fact that they are now being implemented by two
different infrastructures `ExternalStorage` and `KMS`. This is an
expected evolution, and will pave the way for our third and final support
for changefeed `Sinks`. A large portion of this diff is just plumbing the
appropriate environments throught the backup/restore code.

This diff also adds KMS specific tests to `pkg/ccl/cloudccl/gcp` and tweaks
our nightly job to run these tests with the configured credentials.

Informs: https://github.com/cockroachdb/cockroach/issues/84753

Release note (sql change): GCP KMS can be represented as an External Connection
object, that can be used during a backup or restore using the `external` URI.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/85075)
<!-- Reviewable:end -->

adityamaru committed:
externalconn,backupccl: add KMS support to External Connections

In #84931 we taught the `ExternalStorage` infrastructure to
recognize the `external` URI scheme. In this change we do the same
but for the `KMS` infrastructure.

Concretely, a user is now able to create an external storage
object that represent a gcp KMS URI. This can be done using the
`CREATE EXTERNAL CONNECTION` syntax. The user is then able to point
an operation in CockroachDB such as BACKUP,RESTORE,SHOW BACKUP to that
KMS using an `external` URI. For example:

```
CREATE EXTERNAL CONNECTION backup AS 'nodelocal://1/foo';
CREATE EXTERNAL CONNECTION backupkms AS 'gs:///cmk?AUTH=implict';

BACKUP INTO 'external://foo' WITH kms='external://backupkms';
```

Under the hood, we implement the `ConnectionDetails` interface
for gcp KMS. This allows us to persist a row in the `external_connections`
table when the object is created, and to `Dial` the underlying resource
when the object is being used. The interfaces had to be tweaked slightly
to accomodate for the fact that they are now being implemented by two
different infrastructures `ExternalStorage` and `KMS`. This is an
expected evolution, and will pave the way for our third and final support
for changefeed `Sinks`. A large portion of this diff is just plumbing the
appropriate environments throught the backup/restore code.

This diff also adds KMS specific tests to `pkg/ccl/cloudccl/gcp` and tweaks
our nightly job to run these tests with the configured credentials.

Informs: #84753

Release note (sql change): GCP KMS can be represented as an External Connection
object, that can be used during a backup or restore using the `external` URI.

adityamaru commented:
A lot of this diff is plumbing the `kmsEnv` from planning/job execution through the various methods. This is necessary now because we need more things to be present in the `kmsEnv` to be able to talk to the `system.external_connections` table. I would start by focusing on the diffs in `pkg/cloud/externalconn`, `pkg/cloud/gcp` and `pkg/ccl/cloudccl/gcp`.

benbardin commented on pkg/cloud/externalconn/connection.go:
```diff
@@ -22,14 +22,24 @@ import (
+type ConnectionContext interface {
+	// ExternalStorageContext returns the dependencies passed to external storage
+	// implementation during creation.
+	ExternalStorageContext() cloud.ExternalStorageContext
```
This is a oneof situation, right? Is there any way of adding a hard-coding that? Or maybe just add a comment?

adityamaru replied:
Yeah it is expected that only one of these methods is implemented depending on who is `Dial`ing. I wonder if @miretskiy or you have any better ideas of designing this, I can definitely add a comment to that effect.

miretskiy replied:
I was initially thinking of just using types.Any -- but we wound up w/ oneOf...
Types.Any, of course would mean that the implementations of Dial would be responsible of ensuring that
the type they get is the expected type.

benbardin commented on pkg/cloud/externalconn/connection_kms.go:
```diff
@@ -0,0 +1,88 @@
+
+// ExternalStorageContext implements the ConnectionContext interface.
+func (k *kmsConnectionContext) ExternalStorageContext() cloud.ExternalStorageContext {
+	panic("kmsConnectionContext cannot be used for External Storage initialization")
```
Maybe return an err explicitly? This may be to subtle for users, who would probably prefer to be told "This is a KMS external connection, You can't use this as a Storage."

adityamaru replied:
I'm happy adding an error return type, I've been told in the past to keep signatures simpler and rely on panics to code that if reached is clearly a programming error. I'm not worried about it being subtle, it'll crash the cockroach node with the panic if anyone tries to misuse this.

benbardin replied:
Oh, this is programmer error? A user can't get this by trying to use a KMS external conn in a STORAGE external conn slot?

adityamaru replied:
Oh yikes you're right, really good catch. I'm convinced this should be an error, but additionally we should have a check before we attempt to `Dial`. We should make sure that if you're asking for a KMS then the details is actually of type KMS, and similarly for storage. Making the change and adding a test!

adityamaru replied:
I've addressed this in the `LoadConnection` library method. It now checks that the object is of the correct type.

benbardin commented on pkg/cloud/externalconn/connectionpb/connection.proto:
```diff
@@ -18,23 +18,35 @@ import "cloud/cloudpb/external_storage.proto";
+  KMS = 2 [(gogoproto.enumvalue_customname) = "TypeKMS"];
 }
 
 message ConnectionDetails {
```
Having a Details be completely unaware of its `ConnectionType` is starting to feel weird to me (mea culpa for not noticing this earlier) - it just seems kind of strange to have `nodelocal` and `gcs_kms` in the same list, when they're such radically different categories (as indicated by that `kms` suffix!). Think there's value in folding a Type into a ConnectionDetails, and then perhaps having a nested oneof's? I.e. something like
```
oneof detailsByType {
  StorageDetails storage = 1
  KMSDetails ks = 2
}
```

and then in `KMSDetails`: 
```
oneof details {
  GCSKMSConnectionDetails gcs_kms =1
  AWS... aws_kms=2
}
```
Or maybe that's too verbose for not enough gain? Or just wrong? Or difficult? We could also leave as-is, and just rename `nodelocal` to `storage_nodelocal` or some such (either by flat-out renaming, which is probably fine under the circumstances, or by deprecating field 2 and adding the new field as number 4)

adityamaru replied:
The reason for having a `ConnectionType` be an External Connection object level property (and a column in the system table) was to allow for tools like:

`SHOW BACKUP EXTERNAL CONNECTIONS` 

to be built in the future. These could be driven by delegate `SELECT` queries that filter out relevant rows without having to unmarshal each detail and look inside it. As you might've suggested before maybe we could add a method to `connection.go` that looks something like:

`func (d *ConnectionDetails) Type() {}` that returns the type based on the Details we give it. See `pkg/jobs/jobspb/wrap.go` I think jobs does a similar thing. That way we tie details to their types without actually storing them in the details proto. Wdyt?

benbardin replied:
I am convinced that we want a database column with connection type in it!

I'm not convinced that we've nailed a strong polymorphic model here yet. Thinking about it, do we actually the different types of external connections to be the same object? They have a `Dial` API in common...only we don't want that `Dial` method to actually be cross-callable, which feels like the sort of thing a compiler can help with! But each external connection type is created differently, stored differently and used differently...do they have enough in common to justify a shared struct and proto? Very curious what @miretskiy thinks too!

adityamaru replied:
> But each external connection type is created differently, stored differently and used differently  

Isn't this what an interface is solving for us? We have different implementations of this interface that each know what its details are, how they are stored, what to do with them, what to redact and so on. As far as a shared proto goes, my understanding is that a `oneof` is appropriate for this use case where you have many independent protos, each with their own unique properties but belonging to the same umbrella of an `External Connection`. Jobs is similar in its layout where we have a top-level `Details` proto with several specialized `BACKUP, RESTORE,CDC` details nested in it. Each job decides how to implement resume, onfailorcancel etc.

adityamaru replied:
Not being able to call `Dial` on a KMS object when you actually want an ExternalStorage seems like a problem that can be solved by consolidating the codepath that reads these objects from the system table. We can come up with an API where you have to go to the method with a type you want, and the method will only return you the object if it is of that type.

benbardin replied:
Discussed offline!
- We're pretty confident we won't need more than 2/3 connection types for a while, but not certain. That means we want to store all of these in the same database table - while it's scalable enough to add the occasional struct or interface, system tables are more heavyweight.
- Given that all of these external connection types are going in the same table, we need to use some kind of type-checking/casting to convert the row field into the appropriate type of object. This can and will happen exclusively in the external connection library.
- Given _that_, having multiple interfaces doesn't buy us anything. So we probably also want to leave a single top-level proto for consistency.

benbardin reviewed (commented):
Starting with the core struct/proto - will review the rest once we're aligned there!

miretskiy commented on pkg/cloud/externalconn/connection.go:
```diff
@@ -22,14 +22,24 @@ import (
 
+// ConnectionContext contains the dependencies passed to external connection
+// implementation during creation.
+type ConnectionContext interface {
```
does this have to be an interface?
could it be a simple struct?

adityamaru replied:
I think it could be a struct but would we just leave some fields unset depending on where we're Dialing from? For example when passing in a `kmsEnv` we don't want to be setting `limiters` that is exclusively an `ExternalStorage` concept.

benbardin replied:
i'm with aditya, i vote interface!

miretskiy replied:
> i'm with aditya, i vote interface!

Too bad -- it's a configuratoin struct.  That's what it is.
The fact that some configuration may or may not be applicable is not a reason for additional complexity.

benbardin replied:
> > i'm with aditya, i vote interface!
> 
> Too bad -- it's a configuratoin struct. That's what it is. The fact that some configuration may or may not be applicable is not a reason for additional complexity.

The complexity is intrinsic to the problem of multiple-backends-with-a-singular-API, no? We could manage that complexity through a single struct whose fields are used inconsistently. Or, we could use multiple structs with a single consistently-used interface. Why is the former so much better?

adityamaru replied:
fwiw while adding support for a kafka sink I'm leaning towards changing Dial() to take in an `interface{}` because of the CCL/non-CCL boundary and the complications that arise with that. I do expect the signatures to evolve in the next couple of PRs so there will be more chances to tweak it.

benbardin commented on pkg/ccl/backupccl/backup_processor_planning.go:
```diff
@@ -67,13 +67,16 @@ func distBackupPlanSpecs(
 			return nil, err
 		}
+		defer func() {
+			err := kms.Close()
```
Will this error for a non-external-connection KMS?

adityamaru replied:
This doesn't have to do with this diff, this was wrong before external connections existed. I just saw it manifest as leaked goroutines in my test. We should always close a KMS after we're done using it. I actually pulled this fix out into its own PR - https://github.com/cockroachdb/cockroach/pull/85182

benbardin reviewed (approved):
LGTM. As discussed, please also get a review from someone well-versed in KMS. Thank you!

adityamaru commented:
Apologies for the push noise, I'm just trying to get the cloud tests working on our nightly job. I will ping when this is ready for a final look.

adityamaru commented on pkg/ccl/backupccl/alter_backup_planning.go:
```diff
@@ -182,20 +181,17 @@ func doAlterBackupPlan(
 	var plaintextDataKey []byte
-	plaintextDataKey, err = backupencryption.GetEncryptionKey(ctx, encryption, baseStore.Settings(),
-		baseStore.ExternalIOConf())
+	plaintextDataKey, err = backupencryption.GetEncryptionKey(ctx, encryption, &kmsEnv)
```
We're just plumbing the kmsEnv through, no functional change.

adityamaru commented:
Okay this should be ready for a look, cloud unit tests are passing - https://teamcity.cockroachdb.com/viewLog.html?buildId=5912571&buildTypeId=Cockroach_Nightlies_CloudUnitTests&tab=buildLog&_focus=4777#_state=4773,4777

Things that have changed since last push:
- Added more tests around assume role, and trying to use a STORAGE object as a KMS.
- Added a `connection.go` file so that the ConnectionType is a derived property of the details.

@rhu713 I'd appreciate a look at the kms changes, I'm happy to answer any questions about external connections you might have. Roughly `external_connection_kms.go` is the file that has all the logic to interpret an `external://` URI, and `gcs_kms_connection.go` has all the logic that allows an EC object to represent a KMS URI. The `pkg/ccl/cloudccl/gcp` tests use external connections to run backups with gcp KMS so that might be a good place to have a look too.

adityamaru commented on pkg/cloud/externalconn/connectionpb/connection.proto:
```diff
@@ -18,23 +18,35 @@ import "cloud/cloudpb/external_storage.proto";
   cloudpb.LocalFileConfig cfg = 1 [(gogoproto.nullable) = false];
 }
+
+message GCSKMSConnectionDetails {
```
Another thought that crossed my mind while working on kafka support is if we should even be disassembling this URI only to put it back together when it is used. This becomes quite cumbersome when you start having 10s of params like kafka. Why not just stick the URI string in this proto after performing the relevant validation during `CREATE EXTERNAL CONNECTION`. Thoughts @miretskiy @benbardin

miretskiy commented:
I thought that's what the plan was

On Sat, Jul 30, 2022 at 12:27 PM Aditya Maru ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In pkg/cloud/externalconn/connectionpb/connection.proto
> <https://github.com/cockroachdb/cockroach/pull/85075#discussion_r933854939>
> :
>
> >    }
>  }
>
>  message NodelocalConnectionDetails {
>    cloudpb.LocalFileConfig cfg = 1 [(gogoproto.nullable) = false];
>  }
> +
> +message GCSKMSConnectionDetails {
>
> Another thought that crossed my mind while working on kafka support is if
> we should even be disassembling this URI only to put it back together when
> it is used. This becomes quite cumbersome when you start having 10s of
> params like kafka. Why not just stick the URI string in this proto after
> performing the relevant validation during CREATE EXTERNAL CONNECTION.
> Thoughts @miretskiy <https://github.com/miretskiy> @benbardin
> <https://github.com/benbardin>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/85075#pullrequestreview-1056429524>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANA4FVCE4VWQCUMFGGY2MZ3VWVJXPANCNFSM54W43WMQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

rhu713 commented on pkg/cloud/gcp/gcs_kms_connection.go:
```diff
@@ -0,0 +1,73 @@
+func init() {
+	externalconn.RegisterConnectionDetailsFromURIFactory(
+		connectionpb.ConnectionProvider_gs_kms,
+		gcsScheme,
```
I see your point now about how this is conflicting with the naming scheme for GCS external connections whenever that gets added.

rhu713 reviewed (approved):
LGTM!

adityamaru commented:
TFTR! borsing this unless anyone objects, since it makes subsequent reviews easier. If there are areas we want to iterate on we can always revisit them as part of https://github.com/cockroachdb/cockroach/pull/85410 and future PRs.

bors r=benbardin,rhu713
